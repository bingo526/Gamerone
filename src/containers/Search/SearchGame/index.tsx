import React, {
  useState,
  useEffect,
  useCallback,
  useRef,
  useLayoutEffect,
} from 'react';
import { useHistory } from 'react-router-dom';
import DebouncedInput from 'components/common/DebouncedInput';

import GameListItem from '../GameListItem';
import useSearchApi from 'lib/useSearchApi';

// api
import * as GameApi from 'api/game';

import './style.scss';
import Error from 'components/common/Message/Error';
import { Game } from 'interfaces';

export interface SerachGameProps {
  navigate?: boolean;
  multiple?: boolean;
  clearInput?: boolean;
  initLoad?: boolean;
  onAdd?: (id: Game) => void;
  onRemove?: (id: Game) => void;
  limit?: number;
  initGames?: Game[];
}

function SearchGame({
  navigate = true,
  multiple = false,
  clearInput = true,
  initLoad = false,
  onAdd,
  onRemove,
  limit,
  initGames = [],
  ...rest
}: SerachGameProps & React.HTMLProps<HTMLDivElement>) {
  const searchRef = useRef<HTMLInputElement>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);

  const history = useHistory();
  const [query, setQuery] = useState<string | null | undefined>(
    initLoad ? '' : null,
  );
  const [dropdownHidden, setDropdownHidden] = useState(!initLoad);
  const [{ results: games, isSearching, isError }, doFetch] = useSearchApi<
    Game
  >(GameApi.searchGame, initLoad ? '' : null, []);
  const [selectedGames, setSelectedGames] = useState<Game[]>([]);

  useEffect(() => {
    doFetch(query);
    if (query === null) setQuery(undefined);
  }, [query, doFetch]);

  useEffect(() => {
    setSelectedGames(initGames);
  }, [initGames]);

  useLayoutEffect(() => {
    if (searchRef.current) searchRef.current.focus();
  }, [searchRef]);

  const handleInputChange = useCallback(
    (e) => {
      setQuery(e.target.value);
    },
    [setQuery],
  );

  const handleEscKey = useCallback((e) => {
    if (e.keyCode === 27) {
      setDropdownHidden(true);
      if (searchRef.current) searchRef.current.blur();
    }
  }, []);

  const handleDocumentMousedown = useCallback((e) => {
    if (
      dropdownRef.current &&
      !dropdownRef.current.contains(e.target) &&
      searchRef.current &&
      searchRef.current !== e.target
    ) {
      setDropdownHidden(true);
    }
  }, []);

  const handleInputFocus = useCallback(() => {
    setDropdownHidden(false);
  }, []);

  const handleItemSelect = useCallback(
    (game: Game) => {
      if (!navigate) {
        if (onAdd) onAdd(game);

        // unless it surpasses the limit
        if (!limit || selectedGames.length < limit)
          setSelectedGames([...selectedGames, game]);

        // selecting one-item list should hide the dropdown
        if (games.length === 1 && games[0].id === game.id)
          setDropdownHidden(true);
      } else if (navigate) {
        history.push('/' + game.name);
      }

      if (clearInput) {
        setQuery(null);
        setDropdownHidden(true);
        if (searchRef.current) searchRef.current.blur();
      } else if (multiple) {
        if (searchRef.current) searchRef.current.focus();
      } else {
        setDropdownHidden(true);
      }
    },
    [
      games,
      navigate,
      multiple,
      clearInput,
      selectedGames,
      onAdd,
      limit,
      history,
    ],
  );

  const handleItemUnselect = useCallback(
    (game: Game) => {
      setSelectedGames(selectedGames.filter((g) => g.id !== game.id));
      if (onRemove) onRemove(game);
    },
    [selectedGames, onRemove],
  );

  const isBlocked = useCallback(
    (id) => {
      return (
        initGames.findIndex((g) => g.id === id) !== -1 ||
        selectedGames.findIndex((p) => id === p.id) !== -1
      );
    },
    [initGames, selectedGames],
  );

  useEffect(() => {
    document.addEventListener('keydown', handleEscKey, false);
    document.addEventListener('mousedown', handleDocumentMousedown);
    return () => {
      document.removeEventListener('keydown', handleEscKey, false);
      document.removeEventListener('mousedown', handleDocumentMousedown);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <div {...rest}>
      <form className="form-inline" aria-label="search-form">
        <DebouncedInput
          inputMode="text"
          value={query}
          className="input debounced-input"
          placeholder="Search Game"
          aria-label="search-games"
          ref={searchRef}
          onChange={handleInputChange}
          onFocus={handleInputFocus}
          isLoading={isSearching}
        />
      </form>

      {isError && <Error />}

      {!dropdownHidden && games && (
        <div
          className="search-game-dropdown"
          ref={dropdownRef}
          aria-label="search-dropdown"
        >
          {games.map((game: Game) => (
            <GameListItem
              game={game}
              key={game.id}
              handleClick={handleItemSelect}
              blocked={isBlocked(game.id)}
            />
          ))}
        </div>
      )}

      {selectedGames && (
        <div aria-label="selected-games-list">
          {selectedGames.map((game: Game) => (
            <GameListItem
              key={game.id}
              game={game}
              chosen
              handleItemRemove={handleItemUnselect}
            />
          ))}
        </div>
      )}
    </div>
  );
}

export default React.memo(SearchGame);
