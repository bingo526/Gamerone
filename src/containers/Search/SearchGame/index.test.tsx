import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchGame from './index';

const defaultGames = [
  {
    id: 1775,
    name: 'Xenoblade Chronicles X',
    cover: 'games/xenoblade_chronicles_x.jpg',
    logo: null,
    playerCount: 0,
  },
  {
    id: 493,
    name: 'PokÃ©mon X/Y',
    cover: 'games/placeholder.jpg',
    logo: null,
    playerCount: 0,
  },
  {
    id: 1052,
    name: 'Final Fantasy X',
    cover: 'games/final_fantasy_x.jpg',
    logo: null,
    playerCount: 0,
  },
];

const searchGame = async (
  {
    multiple = false,
    navigate = false,
    clearInput = false,
    initLoad = false,
  } = {},
  additions = null,
) => {
  const jestFetch = jest
    .spyOn(window, 'fetch')
    .mockImplementationOnce(async () => {
      return Promise.resolve({
        status: 200,
        json: () => defaultGames,
      });
    });

  const utils = render(
    <MemoryRouter>
      <SearchGame
        multiple={multiple}
        navigate={navigate}
        clearInput={clearInput}
        initLoad={initLoad}
      />
      {additions}
    </MemoryRouter>,
  );

  act(() => {
    fireEvent.change(utils.getByLabelText(/search-games/), {
      target: { value: 'game' },
    });
  });

  await waitForElement(() => utils.getByLabelText(/search-dropdown/));
  await waitForElement(() => utils.getAllByLabelText(/game-list-item/));

  jestFetch.mockRestore();
  return { ...utils, jestFetch };
};

describe('basic functionality', () => {
  test('searches games, shows the dropdown and hides it on outside click', async () => {
    const { asFragment, getByLabelText, queryByLabelText } = await searchGame();
    expect(asFragment()).toMatchSnapshot();

    act(() => {
      fireEvent.mouseDown(getByLabelText(/search-form/));
    });

    expect(queryByLabelText(/search-dropdown/)).toBe(null);
  });

  test('input stays the focused after search', async () => {
    const { getByLabelText } = await searchGame();

    const input = getByLabelText(/search-games/);
    expect(document.activeElement).toBe(input);
  });
});

describe('option: multiple', () => {
  test('adds item to the list on click and not hide the dropdown', async () => {
    const { getByLabelText } = await searchGame({ multiple: true });

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstGame = dropdown.querySelectorAll('.game-list-item')[0];
    act(() => fireEvent.click(firstGame));

    const selectedList = getByLabelText(/selected-games-list/);
    expect(selectedList).toBeInTheDocument();

    const selectedGameName = selectedList.querySelectorAll('.game-name')[0];

    // ensure that same named item is added
    expect(firstGame.querySelector('.game-name').innerHTML).toBe(
      selectedGameName.innerHTML,
    );

    expect(dropdown).toBeInTheDocument();
  });
});

describe('option: navigate', () => {
  test('navigates to the game page on item click', async () => {
    let capturedlocation;
    const { getByLabelText } = await searchGame(
      { navigate: true },
      <Route
        path="*"
        render={({ history, location }) => {
          capturedlocation = location;
          return null;
        }}
      />,
    );

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstGame = dropdown.querySelectorAll('.game-list-item')[0];
    const gameName = firstGame.querySelector('.game-name').innerHTML;
    act(() => fireEvent.click(firstGame));

    expect(capturedlocation.pathname).toBe(`/${gameName}`);
  });
});

describe('option: clearInput', () => {
  test('clears the input after the item is clicked', async () => {
    const { getByLabelText, asFragment } = await searchGame({
      clearInput: true,
    });

    const dropdown = getByLabelText(/search-dropdown/);
    expect(dropdown).toBeInTheDocument();

    const firstGame = dropdown.querySelectorAll('.game-list-item')[0];
    act(() => fireEvent.click(firstGame));

    expect(asFragment()).toMatchSnapshot();
    const input = getByLabelText(/search-games/);

    expect(input.value).toBe('');
  });
});

describe('option: initLoad', () => {
  test('should fetch data and show the dropdown initially', async () => {
    const jestFetch = jest
      .spyOn(window, 'fetch')
      .mockImplementationOnce(async () => {
        return Promise.resolve({
          status: 200,
          json: () => defaultGames,
        });
      });

    const { getAllByLabelText, getByLabelText, asFragment } = render(
      <MemoryRouter>
        <SearchGame initLoad={true} />
      </MemoryRouter>,
    );

    await waitForElement(() => getByLabelText(/search-dropdown/));
    await waitForElement(() => getAllByLabelText(/game-list-item/));

    expect(asFragment()).toMatchSnapshot();

    expect(jestFetch).toBeCalled();
    expect(jestFetch).toHaveBeenCalledTimes(1);

    const items = getAllByLabelText(/game-list-item/);
    expect(items.length).toBe(defaultGames.length);

    jestFetch.mockRestore();
  });

  test('should not fetch data and not show the dropdown on load', async () => {
    const jestFetch = jest
      .spyOn(window, 'fetch')
      .mockImplementationOnce(async () => {
        return Promise.resolve({
          status: 200,
          json: () => defaultGames,
        });
      });

    const { queryAllByLabelText, getByLabelText, asFragment } = render(
      <MemoryRouter>
        <SearchGame initLoad={false} />
      </MemoryRouter>,
    );

    await waitForElement(() => getByLabelText(/search-dropdown/));

    const items = queryAllByLabelText(/game-list-item/);
    expect(items.length).toBe(0);

    expect(asFragment()).toMatchSnapshot();
    expect(jestFetch).toHaveBeenCalledTimes(0);

    jestFetch.mockRestore();
  });
});
