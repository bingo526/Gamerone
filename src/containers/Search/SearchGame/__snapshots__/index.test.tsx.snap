// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality searches games, shows the dropdown and hides it on outside click 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-games"
          class="form-control"
          inputmode="text"
          placeholder="Search Game"
          value="game"
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-game-dropdown"
    >
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Xenoblade Chronicles X"
            id="Xenoblade Chronicles X-cover"
            src="https://cdn.g1.gg/128/games/xenoblade_chronicles_x.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Xenoblade Chronicles X
        </div>
      </div>
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Pokémon X/Y"
            id="Pokémon X/Y-cover"
            src="https://cdn.g1.gg/128/games/placeholder.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Pokémon X/Y
        </div>
      </div>
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Final Fantasy X"
            id="Final Fantasy X-cover"
            src="https://cdn.g1.gg/128/games/final_fantasy_x.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Final Fantasy X
        </div>
      </div>
    </div>
    <div
      aria-label="selected-games-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;

exports[`option: clearInput clears the input after the item is clicked 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-games"
          class="form-control"
          inputmode="text"
          placeholder="Search Game"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="selected-games-list"
      class="d-flex flex-wrap"
    >
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item chosen"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Xenoblade Chronicles X"
            id="Xenoblade Chronicles X-cover"
            src="https://cdn.g1.gg/128/games/xenoblade_chronicles_x.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Xenoblade Chronicles X
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`option: initLoad should fetch data and show the dropdown initially 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-games"
          class="form-control"
          inputmode="text"
          placeholder="Search Game"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-game-dropdown"
    >
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Xenoblade Chronicles X"
            id="Xenoblade Chronicles X-cover"
            src="https://cdn.g1.gg/128/games/xenoblade_chronicles_x.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Xenoblade Chronicles X
        </div>
      </div>
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Pokémon X/Y"
            id="Pokémon X/Y-cover"
            src="https://cdn.g1.gg/128/games/placeholder.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Pokémon X/Y
        </div>
      </div>
      <div
        aria-label="game-list-item"
        class="d-flex game-list-item"
      >
        <div
          class="p-2 game-cover"
        >
          <img
            alt="Final Fantasy X"
            id="Final Fantasy X-cover"
            src="https://cdn.g1.gg/128/games/final_fantasy_x.jpg"
          />
        </div>
        <div
          class="p-2 flex-grow-1 game-name text-wrap text-secondary"
        >
          Final Fantasy X
        </div>
      </div>
    </div>
    <div
      aria-label="selected-games-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;

exports[`option: initLoad should not fetch data and not show the dropdown on load 1`] = `
<DocumentFragment>
  <div>
    <form
      aria-label="search-form"
      class="form-inline"
    >
      <div
        class="position-relative d-block w-100"
      >
        <input
          aria-label="search-games"
          class="form-control"
          inputmode="text"
          placeholder="Search Game"
          value=""
        />
      </div>
    </form>
    <div
      aria-label="search-dropdown"
      class="search-game-dropdown"
    />
    <div
      aria-label="selected-games-list"
      class="d-flex flex-wrap"
    />
  </div>
</DocumentFragment>
`;
